docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,296 @@
+vSphere cluster uses the session as a sanity check to ensure connectivity. vSphere VM uses the session for VM lifecycle tasks on vcenter and the session is stored in a VMContext.
+
+```go
+type VMContext struct {
+    *ControllerContext
+    VSphereVM   *v1alpha3.VSphereVM
+    PatchHelper *patch.Helper
+    Logger      logr.Logger
+    Session     *session.Session
+}
+```
+
+#### Proposed Changes
+
+The proposed changes below allow users to specify the account to use for clusters during runtime.
+
+Changed Resources
+
+- `VSphereCluster`
+
+New Resources
+
+- cluster-scoped `VSphereAccount` represents the vCenter account to use for reconciliation. This type should also contain a string list of namespaces in which it is allowed to be used.
+
+Changes to VSphereCluster
+A new field is added to the VSphereClusterSpec to reference the `VSphereAccount`. We intend to use `corev1.LocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.
+
+```go
+// VSphereAccount is the account to be used for vcenter actions
+type VSphereAccount struct {
+    metav1.TypeMeta `json:`",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec VSphereAccountSpec `json:"spec,omitempty"`
+    Status VSphereAccountStatus `json:"status, omitempty"`
+}
+
+type VSphereAccountSpec struct {
+    Username string `json:"username"`
+
+    Password string `json:"password"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using an array of namespaces or with a label selector.
+    // If this object is nil, no namespaces will be allowed
+    //
+    // +optional
+    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+type AllowedNamespaces struct {
+    // If both NamespaceList and Selector are nil/empty, no namespaces will be allowed to use the account (default behaviour)
+    //
+    // Allowed namespaces listed out by name
+    // +optional
+    // +nullable
+    NamespaceList []string `json:"list"`
+
+    // Selector is a standard Kubernetes LabelSelector. A label query over a set of resources.
+    // +optional
+    Selector metav1.LabelSelector `json:"selector"`
+}
+
+type VSphereClusterSpec struct {
+    ...
+    // +optional
+    AccountRef *corev1.LocalObjectReference `json:"accountRef,omitempty"`
+}
+```
+
+#### Controller Changes
+
+- 1If AccountRef is specified, the CRD is fetched and used to create a session
+- The session will be cached using the existing logic where the key is server + username + datacenter
+- The controller should reconcile the VSphereAccount resource and ensure connectivity to the account
+- The controller will set the VsphereAccount as one of the OwnerReferences of the VsphereCluster if the VsphereAccount is in the same namespace as the VsphereCluster
```

Comment:
I'm a little bit confused by this, my understanding till now was that VsphereAccount is a cluster scoped resource...

Reply:
It should be cluster scoped. CAPA called out cluster scoped vs. namespaced in the proposal but the implementation (so far) uses cluster scoped identities and doesn't set the owner reference. CAPZ seems to use namespaced crds for the AzureIdentity but also copies the identity per cluster.